File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        @cache
        def dfs(i, remain):
            if remain < 0:
                return False
            if i == len(nums):
                return True if remain == 0 else False
            return dfs(i + 1, remain) or dfs(i + 1, remain - nums[i])
        total = sum(nums)
        if total % 2 == 1:
            return False
        return dfs(0, total // 2)
        

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        if (sum % 2 == 1) {
            return false;
        }
        Boolean[][] memo = new Boolean[nums.length][sum / 2 + 1];
        return dfs(0, nums, 0, sum, memo);
    }
    private boolean dfs(int i, int[] nums, int curr, int sum, Boolean[][] memo) {
        if (curr == sum / 2) {
            return true;
        }
        if (i == nums.length || curr > sum / 2) {
            return false;
        }
        if (memo[i][curr] != null) {
            return memo[i][curr];
        }
        memo[i][curr] = dfs(i + 1, nums, curr + nums[i], sum, memo) || dfs(i + 1, nums, curr, sum, memo);
        return memo[i][curr];
    }
}